### DISCUSSION TITLE
Thoughts on migrating to TypeScript?

### DISCUSSION THREAD

**Discussion by fgblomqvist** [resonatecoop/stream#138] [Ideas]
Didn't find any discussions on this so figured I might as well ask. Are there any thoughts on this subject? Naturally it would be a big undertaking but fortunately one that can be done slowly and one-piece-at-a-time thanks to TS having good configuration possibilities regarding that.

I'm not sure if there are any arguments against it other than that it would take some time and energy for the initial setup, along with some slight overhead in terms of maintenance in the long run (but only slight I'd say). Either way, just curious :slightly_smiling_face:

**Comment by peterklingelhofer** [comment]
Personally, I use TypeScript a good bit and would welcome a conversion to TypeScript. Declared typings and interfaces reduce onboarding time for new devs because they don't have to throw a bunch of console.logs throughout the code to know what's going on.

However, I'm not sure if the other devs/maintainers use it or like it though, I would be curious to hear their thoughts on this matter.

**Reply by fgblomqvist** [reply]
Any thoughts on this @auggod?

**Reply by auggod** [reply]
I don't have a strong opinion about typescript although I feel like it does not worth our time. Anyway, I would welcome any new project using typescript at Resonate.

How much time would it take to convert the project to use typescript, even partially?

I don't think we need to convert everything to typescript. What the minimum setup would look like?

**Reply by lindner** [reply]
It's been a while since I've converted but the typescript compiler can handle plain Js and it does catch  a lot of simple errors since it's idiomatic.  You can then choose to move portions of code from js to ts.

https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html

The larger question is how does it fit within the larger build/release ecosystem.  I haven't looked at that or the lerna system for a long long time.  Someone could see what it takes to incorporate typescript.

Personally I think it's the best option on the table for typed JS these days...

**Reply by fgblomqvist** [reply]
A minimal TS setup is generally just a correctly configured tsconfig + the TS dependency + mixing it into your build system. Most build systems support TS out-of-box or using a basic plugin, however, `jalla` does not seem to go that far. I'm not too familiar with the current setup yet, but it looks like `jalla` and `choo` are used together to achieve a SSR setup using the `choo` framework.

The only easy option I see would be to just compile the typescript first, separately using `tsc`, and then run whatever logic already exists. For local development, it would take the shape of something like `tsc --watch` + `jalla serve` in parallel.

As for the monorepo part -- it depends. First of all, I'm not sure if `jalla` does any type of processing to the deps. A lot of builders do not, and their docs don't mention anything. The way we do things right now, we just import the raw source of our monorepo packages straight into `beta`. If we wanted to add a build step to our packages (which we should, it's quite inflexible to write code that goes straight to the browser), we'd just have to update a package to build files into e.g. a `dist` directory (such as `packages/player-component/dist`) and update its package.json accordingly, to declare that we're now exporting the stuff inside `dist` rather than in the root of the package. For local dev, you'd just run `tsc --watch` on all relevant directories (this can be done with a single `tsc` instance, TS has built-in support for these scenarios).

So yeah, it might sound scarier than it is. I could put together an example PR of what a minimal TS setup would look like, setting it up in `beta` + one of the other monorepo packages (without doing any actual TS conversion yet). Won't have time rn but some day. If someone else is eager to do it don't let me stop you either :slightly_smiling_face:

**Reply by fgblomqvist** [reply]
I wanna take a stab at this later this week after I wrap up #153. Have been feeling that not having types have been really slowing down working on that PR.

**Comment by fgblomqvist** [comment]
Threw up a PR with some initial stuff: https://github.com/resonatecoop/stream/pull/160
